---
title: "BCB420 Assignment 2: Differential Gene Expression Analysis and Preliminary ORA"
author: "Veronica Chang"
date: "February 14th, 2023"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    df_print: paged
editor_options: 
  markdown: 
    wrap: sentence
bibliography: a2.bib
---

# Introduction

## Background

## Objectives

## Dataset

## Previous Analysis

## Packages

-   [BiocManager](https://CRAN.R-project.org/package=BiocManager) [@morgan2022]
-   [GEOmetadb](https://bioconductor.org/packages/release/bioc/html/GEOmetadb.html) [@zhu2008]
-   [limma](https://bioconductor.org/packages/release/bioc/html/limma.html) [@smyth2015]
-   [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) [@robinson2010]
-   [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) [@durinck2009]
-   [GEOquery](https://bioconductor.org/packages/release/bioc/html/GEOquery.html) [@davis2007]
-   [Biobase](https://bioconductor.org/packages/release/bioc/html/Biobase.html) [@huber2015]
-   [RColorBrewer](https://cran.r-project.org/package=RColorBrewer) [@neuwirth2022]

```{r, message=FALSE, warning=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")

library("BiocManager")
library("GEOmetadb")
library("limma")
library("edgeR")
library("dplyr")
library("Biobase")
library("biomaRt")
library("GEOquery")
library("ComplexHeatmap")
library("circlize")
library("RColorBrewer")
library("ggplot2")
library("knitr")
library("ggrepel")
library("RSQLite")
library("kableExtra")
```

# Task 1: Differential Gene Expression

## Import data

For our analysis in A1, we saved our data output to a .csv file.
We will need to load the data file to use in our analysis for this assignment.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Load the data file
counts <- read.csv(file = "~/projects/expDataNorm.csv", header = TRUE, row.names = 1)

# View some of the data
head(counts)

# Also loaded in the samples categories into a table
expSamples <- read.csv("~/projects/expSamples.csv", header = TRUE)
```

**Table 1.** First few rows of a table of the normalized data counts of the chosen gene set.
There are `r nrow(counts)` genes from `r ncol(counts)` samples.
Each gene has a unique HGNC symbol identifier.
The samples can be divided into (1) `patient_group`, MS and CON, where MS includes the multiple sclerosis donor samples and CON is the non-neurological control donors, (2) `tissue`, GM and WM, where GM includes grey matter samples and WM includes white matter samples.

## Design model

We will be using the package `edgeR` for this assignment.
edgeR uses a generalized linear model to test for differential expression of RNA-seq expression profiles.
We need to consider how there are two major factors in our data set - `patient_group` and `tissue`.
Ideally, we would account for both in our design model.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Define the groups 
groups <- paste(expSamples$patient_group, expSamples$tissue,sep = ".")
groups <- factor(groups)

# Create design with groups 
modelDesign <- model.matrix(~0 + groups)
DT::datatable(modelDesign[order(gsub("M[0-9]{2}\\.", "", rownames(modelDesign))), ], 
              rownames = TRUE,
              options = list(pageLength = 5, scrollX = TRUE)) %>%
              DT::formatStyle(names(modelDesign), fontSize = "10px")

```

**Table 2.** Differential expression analysis model design, includes both `patient_group` and `tissue` as factors.

Since we are using `edgeR` for downstream analysis, we will first need to create a DGEList object, this will hold the dataset that will be analyzed by edgeR.
We will also evaluate the dispersion.
We will be using a Quasi-likelihood model as our dataset is of RNA-Seq data and Quasi-likelihood models are best suited.
However, Quasi-likelihood models require that the data follows a negative binomial distribution so will want to confirm whether our data set follows this.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Create DGEList
dgeCounts <- edgeR::DGEList(counts = counts, group = groups)

# Check dispersion 
dgeCounts <- edgeR::estimateDisp(dgeCounts, modelDesign)

edgeR::plotMeanVar(dgeCounts,
                   show.raw.vars = TRUE,
                   show.tagwise.vars = TRUE,
                   NBline = TRUE,
                   show.ave.raw.vars = TRUE,
                   show.binned.common.disp.vars = TRUE,
                   main = "Mean-Variance Plot for Normalized Data")
```

**Figure 1.** Mean-variance relationship of normalized data.
Dispersion estimation is done with the negative binomial model, which estimates the common, trended and tagwise dispersions across all tags.
The grey and light blue points show the mean and pooled variance for each gene.
The red crosses show binned variance, in which genes are grouped by mean level.
The blue line indicates the common dispersion and is based on the negative binomial model.
The black line displays the theoretical variance under the Poisson distribution model where the variance is equal to the mean.

In order to make accurate inferences about differential expression, we need to appropriately model the mean-variance relationship in DGEs.
As we can see in Fig.
1, our normalized data set follows the negative binomial distribution.
We can now verify that we can use a Quasi-likelihood model and fit our data to our model.

## Fit the data

```{r message = FALSE, tidy=TRUE, warning=FALSE}
fit <- edgeR::glmQLFit(dgeCounts, modelDesign)
```

We have now fit our model and can move to differential expression analysis.

## MS vs CON

We will be testing for differential expression between the MS and CON samples.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Create a contrast for the patient groups GM
contrast_gm <- limma::makeContrasts("groupsMS.GM - groupsCON.GM", 
                                        levels = modelDesign)

# Create a contrast for the patient groups WM
contrast_wm <- limma::makeContrasts("groupsMS.WM - groupsCON.WM", 
                                        levels = modelDesign)

# Perform DE analysis
qlf_gm <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_gm)

qlf_wm <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_wm)

# Extract the top DE hits ranked by p-value
results_gm <- edgeR::topTags(qlf_gm, sort.by = "PValue", 
                                   n = nrow(dgeCounts))$table

results_wm <- edgeR::topTags(qlf_gm, sort.by = "PValue", 
                                   n = nrow(dgeCounts))$table
```

We can examine the DE genes pass the threshold and correction. We are using 0.05 as the threshold for the p-value by convention.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of unadjusted p-value < 0.05
sum(results_gm$PValue < 0.05)
sum(results_wm$PValue < 0.05)

# Check if there are any differences between the two tissues
check <- results_wm == results_gm
which(check == "FALSE")
```

The two results are actually the same. Will keep this in mind. We actually have a lot of DE genes with p-value < 0.05, we might want to consider bringing this p-value threshold down to 0.01.


```{r tn_pval_stringent, message=FALSE}
sum(results_gm$PValue < 0.01)
sum(results_wm$PValue < 0.01)
```

## GM vs WM

We will be testing for differential expression between the different tissues. 

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Create a contrast for the MS GM and WM
contrast_ms <- limma::makeContrasts("groupsMS.GM - groupsMS.WM", 
                                        levels = modelDesign)

# Create a contrast for the CON GM and WM
contrast_con <- limma::makeContrasts("groupsCON.GM - groupsCON.WM", 
                                        levels = modelDesign)

# Perform DE analysis
qlf_ms <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_ms)

qlf_con <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_con)

# Extract the top DE hits ranked by p-value
results_ms <- edgeR::topTags(qlf_ms, sort.by = "PValue", 
                                   n = nrow(dgeCounts))$table

results_con <- edgeR::topTags(qlf_con, sort.by = "PValue", 
                                   n = nrow(dgeCounts))$table
```

We can examine the DE genes pass the threshold and correction. We are using 0.05 as the threshold for the p-value by convention.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of unadjusted p-value < 0.05
sum(results_ms$PValue < 0.05)
sum(results_con$PValue < 0.05)
```

Interestingly, there are differences between the tissues for each patient group. There are many genes so we will also bring the p-value threshold down to 0.01 for this comparison. 

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of unadjusted p-value < 0.01
sum(results_ms$PValue < 0.01)
sum(results_con$PValue < 0.01)
```

Since there are many genes for both comparisons, we will also be filtering out genes based on log fold change (log-FC) later in this analysis.

### Volcano plot MS vs CON

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Label genes
results_gm <- results_gm %>%
    dplyr::mutate(Expression = dplyr::case_when(
                        logFC > 0 & FDR < 0.05 ~ "Up-regulated",
                        logFC < 0 & FDR < 0.05 ~ "Down-regulated",
                        TRUE ~ "Not DE"))

# Plot volcano plot
ggplot2::ggplot(results_gm, aes(logFC, -log(FDR, 10))) +
    ggplot2::geom_point(aes(color = Expression), size = 2/5) +
    xlab(expression("log"[2]*"FC")) + 
    ylab(expression("-log"[10]*"FDR")) + 
    ggplot2::scale_color_manual(values = c("Up-regulated" = "firebrick3", 
                                           "Down-regulated" = "dodgerblue3", 
                                           "Not DE" = "gray50")) +
    ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size=1.5))) +
    ggrepel::geom_label_repel(data = results_gm, 
                        mapping = aes(logFC, -log(FDR, 10), label = rownames(results_gm)),
                        size = 2.5, color = "black", nudge_x = 0.1, nudge_y = 0.1)


# Label genes
results_wm <- results_wm %>%
    dplyr::mutate(Expression = dplyr::case_when(
                        logFC > 0 & FDR < 0.05 ~ "Up-regulated",
                        logFC < 0 & FDR < 0.05 ~ "Down-regulated",
                        TRUE ~ "Not DE"))

# Plot volcano plot
ggplot2::ggplot(results_wm, aes(logFC, -log(FDR, 10))) +
    ggplot2::geom_point(aes(color = Expression), size = 2/5) +
    xlab(expression("log"[2]*"FC")) + 
    ylab(expression("-log"[10]*"FDR")) + 
    ggplot2::scale_color_manual(values = c("Up-regulated" = "firebrick3", 
                                           "Down-regulated" = "dodgerblue3", 
                                           "Not DE" = "gray50")) +
    ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size=1.5))) +
    ggrepel::geom_label_repel(data = results_wm, 
                        mapping = aes(logFC, -log(FDR, 10), label = rownames(results_wm)),
                        size = 2.5, color = "black", nudge_x = 0.1, nudge_y = 0.1)

```

There are only a few DE genes and all of them are up-regulated in MS patients for each tissue. The gene with the greatest logFC is `ELK2AP` with a logFC of `10.7060661`.

### Volcano plot GM vs WM

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Label genes
results_ms <- results_ms %>%
    dplyr::mutate(Expression = dplyr::case_when(
                        logFC > 0 & FDR < 0.05 ~ "Up-regulated",
                        logFC < 0 & FDR < 0.05 ~ "Down-regulated",
                        TRUE ~ "Not DE"))

# Plot volcano plot
ggplot2::ggplot(results_ms, aes(logFC, -log(FDR, 10))) +
    ggplot2::geom_point(aes(color = Expression), size = 2/5) +
    xlab(expression("log"[2]*"FC")) + 
    ylab(expression("-log"[10]*"FDR")) + 
    ggplot2::scale_color_manual(values = c("Up-regulated" = "firebrick3", 
                                           "Down-regulated" = "dodgerblue3", 
                                           "Not DE" = "gray50")) +
    ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size=1.5))) +
    ggrepel::geom_label_repel(data = results_ms, 
                        mapping = aes(logFC, -log(FDR, 10), label = rownames(results_ms)),
                        size = 2.5, color = "black", nudge_x = 0.1, nudge_y = 0.1)
legend("topright", legend=c("Downregulated in MS","Upregulated in MS", "Not significant"),fill = c("dodgerblue3", "firebrick3", "gray50"), cex = 0.5)


# Label genes
results_con <- results_con %>%
    dplyr::mutate(Expression = dplyr::case_when(
                        logFC > 0 & FDR < 0.05 ~ "Up-regulated",
                        logFC < 0 & FDR < 0.05 ~ "Down-regulated",
                        TRUE ~ "Not DE"))

# Plot volcano plot
ggplot2::ggplot(results_wm, aes(logFC, -log(FDR, 10))) +
    ggplot2::geom_point(aes(color = Expression), size = 2/5) +
    xlab(expression("log"[2]*"FC")) + 
    ylab(expression("-log"[10]*"FDR")) + 
    ggplot2::scale_color_manual(values = c("Up-regulated" = "firebrick3", 
                                           "Down-regulated" = "dodgerblue3", 
                                           "Not DE" = "gray50")) +
    ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size=1.5))) +
    ggrepel::geom_label_repel(data = results_con, 
                        mapping = aes(logFC, -log(FDR, 10), label = rownames(results_con)),
                        size = 2.5, color = "black", nudge_x = 0.1, nudge_y = 0.1)
legend("topright", legend=c("Downregulated in MS","Upregulated in MS", "Not significant"),fill = c("dodgerblue3", "firebrick3", "gray50"), cex = 0.5)

```

We now want to also adjust the p-values using the Benjamini-Hochberg method as we are doing multiple comparisons. This method assumes that the power to detect differential expression is equally likely among all the tests - reducing the false discovery rate.


## Global gene expression

We can now generate a heat map, this is a data graph that will allow us to visualize global expression change over many samples.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# First, as we have two groups for two factors, annotating the heat map will make it easier to read. 
# In A1, I already defined a colour scheme for the groups, as below:
colTissue <- as.factor(expSamples$tissue)
colTissue <- c("grey", "white")[colTissue]
colPatient <- as.factor(expSamples$patient_group)
colPatient <- c("forestgreen", "dodgerblue")[colPatient]

# This is another one for a combination of tissue and patient_group, group
colGroup <- as.factor(expSamples$group)
levels(colGroup) <-  brewer.pal(nlevels(colGroup), "Spectral")
colGroup <- as.character(colGroup)

# Creating an annotation data frame
dfAnnotation <- data.frame(tissue = countsMatrix$tissue,
                      patient_group = countsMatrix$patient_group)
colAnnotation <- list(tissue = colTissue, patient_group = colPatient)

# Heat map annotation
hmapAnnotation <- ComplexHeatmap::HeatmapAnnotation(
                    df = dfAnnotation,
                    col = colAnnotation)

```
