---
title: "BCB420 Assignment 2: Differential Gene Expression Analysis and Preliminary ORA"
author: "Veronica Chang"
date: "March 14th, 2023"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    df_print: paged
editor_options: 
  markdown: 
    wrap: sentence
bibliography: a2.bib
---

# Introduction

## Background

## Objectives

## Dataset

## Previous Analysis

## Packages

-   [BiocManager](https://CRAN.R-project.org/package=BiocManager) [@morgan2022]
-   [GEOmetadb](https://bioconductor.org/packages/release/bioc/html/GEOmetadb.html) [@zhu2008]
-   [limma](https://bioconductor.org/packages/release/bioc/html/limma.html) [@smyth2015]
-   [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) [@robinson2010]
-   [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) [@durinck2009]
-   [GEOquery](https://bioconductor.org/packages/release/bioc/html/GEOquery.html) [@davis2007]
-   [Biobase](https://bioconductor.org/packages/release/bioc/html/Biobase.html) [@huber2015]
-   [RColorBrewer](https://cran.r-project.org/package=RColorBrewer) [@neuwirth2022]

```{r, message=FALSE, warning=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")

library("BiocManager")
library("GEOmetadb")
library("limma")
library("edgeR")
library("dplyr")
library("Biobase")
library("biomaRt")
library("GEOquery")
library("ComplexHeatmap")
library("circlize")
library("RColorBrewer")
library("ggplot2")
library("knitr")
library("ggrepel")
library("RSQLite")
library("kableExtra")
```

# Task 1: Differential Gene Expression

## Import data

For our analysis in A1, we saved our data output to a .csv file.
We will need to load the data file to use in our analysis for this assignment.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Load the data file
counts <- read.csv(file = "~/projects/expDataNorm.csv", header = TRUE, row.names = 1)

# View some of the data
head(counts)

# Also loaded in the samples categories into a table
expSamples <- read.csv("~/projects/expSamples.csv", header = TRUE)
```

**Table 1.** First few rows of a table of the normalized data counts of the chosen gene set.
There are `r nrow(counts)` genes from `r ncol(counts)` samples.
Each gene has a unique HGNC symbol identifier.
The samples can be divided into (1) `patient_group`, MS and CON, where MS includes the multiple sclerosis donor samples and CON is the non-neurological control donors, (2) `tissue`, GM and WM, where GM includes grey matter samples and WM includes white matter samples.

## Design model

We will be using the package `edgeR` for this assignment.
edgeR uses a generalized linear model to test for differential expression of RNA-seq expression profiles.
We need to consider how there are two major factors in our data set - `patient_group` and `tissue`.
Ideally, we would account for both in our design model. We already defined a group that has both of these factors in A1. 

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Define the groups 
groups <- paste(expSamples$group)
groups <- factor(groups)

# Create design with groups 
modelDesign <- model.matrix(~0 + groups)
colnames(modelDesign) <- c("CON.GM", "CON.WM", "MS.GM", "MS.WM")
DT::datatable(modelDesign[order(gsub("M[0-9]{2}\\.", "", rownames(modelDesign))), ], 
              rownames = TRUE,
              options = list(pageLength = 5, scrollX = TRUE)) %>%
              DT::formatStyle(names(modelDesign), fontSize = "10px")

```

**Table 2.** Differential expression analysis model design, includes both `patient_group` and `tissue` as factors.

Since we are using `edgeR` for downstream analysis, we will first need to create a DGEList object, this will hold the dataset that will be analyzed by edgeR.
We will also evaluate the dispersion.
We will be using a Quasi-likelihood model as our dataset is of RNA-Seq data and Quasi-likelihood models are best suited.
However, Quasi-likelihood models require that the data follows a negative binomial distribution so will want to confirm whether our data set follows this.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Create DGEList
dgeCounts <- edgeR::DGEList(counts = counts, group = groups)

# Remove lowly expressed genes
keep.exprs <- filterByExpr(dgeCounts, group = groups)
dgeCounts <- dgeCounts[keep.exprs,, keep.lib.sizes=FALSE]
dgeCounts <- edgeR::calcNormFactors(dgeCounts)

# Check dispersion 
dgeCounts <- edgeR::estimateDisp(dgeCounts, modelDesign)

edgeR::plotMeanVar(dgeCounts,
                   show.raw.vars = TRUE,
                   show.tagwise.vars = TRUE,
                   NBline = TRUE,
                   show.ave.raw.vars = TRUE,
                   show.binned.common.disp.vars = TRUE,
                   main = "Mean-Variance Plot for Normalized Data")
```

**Figure 1.** Mean-variance relationship of normalized data.
Dispersion estimation is done with the negative binomial model, which estimates the common, trended and tagwise dispersions across all tags.
The grey and light blue points show the mean and pooled variance for each gene.
The red crosses show binned variance, in which genes are grouped by mean level.
The blue line indicates the common dispersion and is based on the negative binomial model.
The black line displays the theoretical variance under the Poisson distribution model where the variance is equal to the mean.

In order to make accurate inferences about differential expression, we need to appropriately model the mean-variance relationship in DEGs.
As we can see in Fig.
1, our normalized data set follows the negative binomial distribution.
We can now verify that we can use a Quasi-likelihood model and fit our data to our model.

## Fit the data

```{r message = FALSE, tidy=TRUE, warning=FALSE}
fit <- edgeR::glmQLFit(dgeCounts, modelDesign)
```

We have now fit our model and can move to differential expression analysis.

## MS vs CON

We will be testing for differential expression between the MS and CON samples for each tissue. We want to also adjust the p-values using the Benjamini-Hochberg method as we are doing multiple comparisons. This method assumes that the power to detect differential expression is equally likely among all the tests - reducing the false discovery rate.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Create a contrast for the patient groups GM
contrast_gm <- limma::makeContrasts("MS.GM - CON.GM", 
                                        levels = modelDesign)

# Create a contrast for the patient groups WM
contrast_wm <- limma::makeContrasts("MS.WM - CON.WM", 
                                        levels = modelDesign)

# Perform DE analysis
qlf_gm <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_gm)

qlf_wm <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_wm)

# Extract the top DE hits ranked by p-value
results_gm <- edgeR::topTags(qlf_gm, sort.by = "PValue", adjust.method = "BH", 
                                   n = nrow(dgeCounts))$table

results_wm <- edgeR::topTags(qlf_wm, sort.by = "PValue", adjust.method = "BH",
                                   n = nrow(dgeCounts))$table
```

We can examine the DE genes pass the threshold and correction.
We are using 0.05 as the threshold for the p-value by convention.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of unadjusted p-value < 0.05
sum(results_gm$PValue < 0.05)
sum(results_wm$PValue < 0.05)

```

We actually have a lot of DE genes with p-value \< 0.05, we might want to consider bringing this p-value threshold down to 0.01.

```{r tn_pval_stringent, message=FALSE}
sum(results_gm$PValue < 0.01)
sum(results_wm$PValue < 0.01)
```

We can see that we have `113` DE genes in the grey matter samples between `CON` and `MS.` We have `176` genes in the white matter samples between `CON` and `MS.`

We should now look at FDR as a "true" indicator of differential expression.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of p-value < 0.05
sum(results_gm$FDR < 0.05)
sum(results_wm$FDR < 0.05)
```
Interestingly, there are not many DE genes for the GM comparison - only 4 and all of them are up-regulated. For the WM comparison, there are none. This makes sense as in the MDS we obtained from A1, we saw that the samples cluster together by tissue despite the patient group. From now on, our analysis will be focused on comparing the GM and WM in the two patient groups. 

## GM vs WM

We will be testing for differential expression between the different tissues for each patient group. We want to also adjust the p-values using the Benjamini-Hochberg method as we are doing multiple comparisons. This method assumes that the power to detect differential expression is equally likely among all the tests - reducing the false discovery rate.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Create a contrast for the MS GM and WM
contrast_ms <- limma::makeContrasts("MS.GM - MS.WM", 
                                        levels = modelDesign)

# Create a contrast for the CON GM and WM
contrast_con <- limma::makeContrasts("CON.GM - CON.WM", 
                                        levels = modelDesign)

# Perform DE analysis
qlf_ms <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_ms)

qlf_con <- edgeR::glmQLFTest(glmfit = fit, 
                                  contrast = contrast_con)

# Extract the top DE hits ranked by p-value
results_ms <- edgeR::topTags(qlf_ms, sort.by = "PValue", adjust.method = "BH", 
                                   n = nrow(dgeCounts))$table

results_con <- edgeR::topTags(qlf_con, sort.by = "PValue", adjust.method = "BH", 
                                   n = nrow(dgeCounts))$table
```

We can examine the DE genes pass the threshold and correction.
We are using 0.05 as the threshold for the p-value by convention.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of p-value < 0.05
sum(results_ms$PValue < 0.05)
sum(results_con$PValue < 0.05)
```

Interestingly, there are differences between the tissues for each patient group.
There are many genes so we will also bring the p-value threshold down to 0.01 for this comparison.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of p-value < 0.01
sum(results_ms$PValue < 0.01)
sum(results_con$PValue < 0.01)
```

We should now look at FDR as a "true" indicator of differential expression.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# DE genes of p-value < 0.05
sum(results_ms$FDR < 0.05)
sum(results_con$FDR < 0.05)

# DE genes of p-value < 0.01
sum(results_ms$FDR < 0.01)
sum(results_con$FDR < 0.01)
```

## Volcano plot GM vs WM

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Label genes
results_ms <- results_ms %>%
    dplyr::mutate(Expression = dplyr::case_when(
                        logFC > 0 & FDR < 0.01 ~ "Up-regulated",
                        logFC < 0 & FDR < 0.01 ~ "Down-regulated",
                        TRUE ~ "Not DE"))

# Plot volcano plot
ggplot2::ggplot(results_ms, aes(logFC, -log(FDR, 10))) +
    ggplot2::geom_point(aes(color = Expression), size = 2/5) +
    xlab(expression("log"[2]*"FC")) + 
    ylab(expression("-log"[10]*"FDR")) + 
    ggplot2::scale_color_manual(values = c("Up-regulated" = "firebrick3", 
                                           "Down-regulated" = "dodgerblue3", 
                                           "Not DE" = "gray50")) +
    ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size=1.5))) +
    ggrepel::geom_label_repel(data = results_ms, 
                        mapping = aes(logFC, -log(FDR, 10), label = rownames(results_ms)),
                        size = 2.5, color = "black", nudge_x = 0.1, nudge_y = 0.1)


# Label genes
results_con <- results_con %>%
    dplyr::mutate(Expression = dplyr::case_when(
                        logFC > 0 & FDR < 0.01 ~ "Up-regulated",
                        logFC < 0 & FDR < 0.01 ~ "Down-regulated",
                        TRUE ~ "Not DE"))

# Plot volcano plot
ggplot2::ggplot(results_con, aes(logFC, -log(FDR, 10))) +
    ggplot2::geom_point(aes(color = Expression), size = 2/5) +
    xlab(expression("log"[2]*"FC")) + 
    ylab(expression("-log"[10]*"FDR")) + 
    ggplot2::scale_color_manual(values = c("Up-regulated" = "firebrick3", 
                                           "Down-regulated" = "dodgerblue3", 
                                           "Not DE" = "gray50")) +
    ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size=1.5))) +
    ggrepel::geom_label_repel(data = results_con, 
                        mapping = aes(logFC, -log(FDR, 10), label = rownames(results_con)),
                        size = 2.5, color = "black", nudge_x = 0.1, nudge_y = 0.1)

```

**Figure 2.** Volcano plot of differentially expressed genes comparing **(A)** grey matter and white matter tissues in MS patients, and **(B)** grey matter and white matter tissues in CON patients. Differentially expressed genes are coloured based on whether they are up- or down-regulated. Genes were selected when FDR < 0.05 AND logFC > 0 for up-regulated OR logFC < 0 for down-regulated. If they do not fulfill these requirements, they are labelled as not differentially expressed (DE).

For the MS patients, the most up-regulated gene was `SNAP25` with a logFC of `6.482862` and the most down-regulated gene was `CNTN2` with a logFC of `-6.641904`. For the CON patients, the most up-regulated gene was `FPR2` with a logFC of `3.2941815` and the most down-regulated gene was `CFAP43` with a logFC of `-7.137068`.

## Global gene expression

We can better visualize global expression change over our samples with a heat map. In this heatmap, all the groups - CON GM, CON WM, MS GM, MS WM will be compared against each other. 

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Create a heat map matrix to scale counts to CPM
dgeCounts <- edgeR::calcNormFactors(dgeCounts)
dgeCountsAll <- edgeR::cpm(dgeCounts)

hmapMatrix <- dgeCountsAll
rownames(hmapMatrix) <- rownames(dgeCountsAll)
colnames(hmapMatrix) <- colnames(dgeCountsAll)
hmapMatrix <-  dgeCountsAll %>% t() %>% scale() %>% t()

# In A1, I already defined a colour annotation for the groups, as below:
annoCol <- brewer.pal(4, "Spectral")
names(annoCol) <- c("CON GM", "CON WM", "MS GM", "MS WM")
annoCol <- list(group = annoCol)

# Make annotation dataframe 
annoDF <- data.frame(group = expSamples$group)

# Create the heat map annotation
hmapAnno <- ComplexHeatmap::HeatmapAnnotation(
                    df = annoDF,
                    col = annoCol)

# Check range of heat map matrix and set colour
hmapRange <- range(hmapMatrix)
hmapCol <- circlize::colorRamp2(c(hmapRange[1], 0, hmapRange[2]), 
                                 c("blue", "white", "red"))

# Create heat map
hmapDEG <- ComplexHeatmap::Heatmap(hmapMatrix,
                                    show_column_dend = FALSE,
                                    show_row_dend = FALSE,
                                    show_column_names = TRUE,
                                    show_row_names = FALSE,
                                    row_names_gp = gpar(cex = 0.5),
                                    cluster_rows = TRUE,
                                    column_order = sort(colnames(hmapMatrix)),
                                    col = hmapCol,
                                    top_annotation = hmapAnno) 

draw(hmapDEG,
   column_title="Heat map of top DE genes",
   column_title_gp=grid::gpar(fontsize=12))

```

**Figure 3.** Heat map of global gene expression across all samples. Samples were annotated based on their patient group and tissue on top of each column, its legend is found on the right. The heat map was coloured based on expression of each gene, where red is up-regulated and blue is down-regulated. Values were converted from counts to CPM and scaled. The samples were also hierarchically clustered by their expression (dendrograms for both rows and columns were excluded to keep the figure neat).

## Differential expression analysis Discussion

### 1. Calculate p-values for each of the genes in your expression set. How many genes were significantly differentially expressed? What thresholds did you use and why?



### 2. Multiple hypothesis testing - correct your p-values using a multiple hypothesis correction method. Which method did you use? And Why? How many genes passed correction?



### 3. Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.



### 4. Visualize your top hits using a heatmap. Do your conditions cluster together? Explain why or why not.




# Task 2: Thresholded Overrepresentation Analysis using g:Profiler
For this part of the assignment, we will be doing a thresholded over-representation analysis with g:Profiler. 

## MS GM vs WM
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Define the groups 
msUpregulated <- results_ms[results_ms$logFC > 0 & results_ms$PValue < 0.01, ]
msDownregulated <- results_ms[results_ms$logFC < 0 & results_ms$PValue < 0.01, ]
```
To perform the gene enrichment analysis, we will use the R package for g:Profiler. We used FDR for correction. In our previous course journal entry assignment, we used GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP) - we will be using these sources in this section.

We will be performing the analysis for up-regulated genes and down-regulated genes separately. 

### Up-regulated
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Up-regulated
msUpTopTerms <- gprofiler2::gost(query = rownames(msUpregulated), 
                                  organism = "hsapiens", 
                                  exclude_iea = TRUE,
                                  correction_method = "fdr",
                                  sources = c("GO:BP", "REAC", "WP"))

# Limit term size
msUpTopTerms <- data.frame(
  term_name = msUpTopTerms$result$term_name[msUpTopTerms$result$term_size < 500 &
                                               msUpTopTerms$result$term_size > 1],
  term_id = msUpTopTerms$result$term_id[msUpTopTerms$result$term_size < 500 &
                                           msUpTopTerms$result$term_size > 1],
  source = msUpTopTerms$result$source[msUpTopTerms$result$term_size < 500 &
                                         msUpTopTerms$result$term_size > 1]
)

# Visualize our results
knitr::kable(head(msUpTopTerms, 10), format = "html")
```
**Table 2.** The first top 10 terms across all the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

Number of terms:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
length(msUpTopTerms$term_name)
```

We are going to want to look at the top terms for each source eventually, so we can look at it now. 
```{r message = FALSE, tidy=TRUE, warning=FALSE}
knitr::kable(rbind(msUpTopTerms[msUpTopTerms$source == "GO:BP",][1,],
                   msUpTopTerms[msUpTopTerms$source == "REAC",][1,],
                   msUpTopTerms[msUpTopTerms$source == "WP",][1,]),
             format = "html")
```
**Table 3.** The top term from each of the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

### Down-regulated
Now, do the same for the down-regulated genes.
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Down-regulated
msDownTopTerms <- gprofiler2::gost(query = rownames(msDownregulated), 
                                  organism = "hsapiens", 
                                  exclude_iea = TRUE,
                                  correction_method = "fdr",
                                  sources = c("GO:BP", "REAC", "WP"))

# Limit term size
msDownTopTerms <- data.frame(
  term_name = msDownTopTerms$result$term_name[msDownTopTerms$result$term_size < 500 &
                                               msDownTopTerms$result$term_size > 1],
  term_id = msDownTopTerms$result$term_id[msDownTopTerms$result$term_size < 500 &
                                           msDownTopTerms$result$term_size > 1],
  source = msDownTopTerms$result$source[msDownTopTerms$result$term_size < 500 &
                                         msDownTopTerms$result$term_size > 1]
)

# Visualize our results
knitr::kable(head(msDownTopTerms, 10), format = "html")
```
**Table 4.** The first top 10 terms across all the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

Number of terms:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
length(msDownTopTerms$term_name)
```

We are going to want to look at the top terms for each source eventually, so we can look at it now. 
```{r message = FALSE, tidy=TRUE, warning=FALSE}
knitr::kable(rbind(msDownTopTerms[msDownTopTerms$source == "GO:BP",][1,],
                   msDownTopTerms[msDownTopTerms$source == "REAC",][1,],
                   msDownTopTerms[msDownTopTerms$source == "WP",][1,]),
             format = "html")
```
**Table 5.** The top term from each of the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 


### All DE genes
We will now be looking at all the differentially expressed genes as a whole. 
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Up-regulated
msAllTopTerms <- gprofiler2::gost(query = rownames(results_ms), 
                                  organism = "hsapiens", 
                                  exclude_iea = TRUE,
                                  correction_method = "fdr",
                                  sources = c("GO:BP", "REAC", "WP"))

# Limit term size
msAllTopTerms <- data.frame(
  term_name = msAllTopTerms$result$term_name[msAllTopTerms$result$term_size < 500 &
                                               msAllTopTerms$result$term_size > 1],
  term_id = msAllTopTerms$result$term_id[msAllTopTerms$result$term_size < 500 &
                                           msAllTopTerms$result$term_size > 1],
  source = msAllTopTerms$result$source[msAllTopTerms$result$term_size < 500 &
                                         msAllTopTerms$result$term_size > 1]
)

# Visualize our results
knitr::kable(head(msAllTopTerms, 10), format = "html")
```
**Table 6.** The first top 10 terms across all the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

Number of terms:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
length(msAllTopTerms$term_name)
```

Look for the top terms from each source:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
knitr::kable(rbind(msAllTopTerms[msAllTopTerms$source == "GO:BP",][1,],
                   msAllTopTerms[msAllTopTerms$source == "REAC",][1,],
                   msAllTopTerms[msAllTopTerms$source == "WP",][1,]),
             format = "html")
```
**Table 7.** The top term from each of the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

## CON GM vs WM
We will be doing the same analysis for the CON patient group.
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Define the groups 
conUpregulated <- results_con[results_con$logFC > 0 & results_con$PValue < 0.01, ]
conDownregulated <- results_con[results_con$logFC < 0 & results_con$PValue < 0.01, ]
```
To perform the gene enrichment analysis, we will use the R package for g:Profiler. We used FDR for correction. In our previous course journal entry assignment, we used GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP) - we will be using these sources in this section.

We will be performing the analysis for up-regulated genes and down-regulated genes separately. 

### Up-regulated
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Up-regulated
conUpTopTerms <- gprofiler2::gost(query = rownames(conUpregulated), 
                                  organism = "hsapiens", 
                                  exclude_iea = TRUE,
                                  correction_method = "fdr",
                                  sources = c("GO:BP", "REAC", "WP"))

# Limit term size
conUpTopTerms <- data.frame(
  term_name = conUpTopTerms$result$term_name[conUpTopTerms$result$term_size < 500 &
                                               conUpTopTerms$result$term_size > 1],
  term_id = conUpTopTerms$result$term_id[conUpTopTerms$result$term_size < 500 &
                                           conUpTopTerms$result$term_size > 1],
  source = conUpTopTerms$result$source[conUpTopTerms$result$term_size < 500 &
                                         conUpTopTerms$result$term_size > 1]
)

# Visualize our results
knitr::kable(head(conUpTopTerms, 10), format = "html")
```
**Table 8.** The first top 10 terms across all the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

Number of terms:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
length(conUpTopTerms$term_name)
```

We are going to want to look at the top terms for each source eventually, so we can look at it now. 
```{r message = FALSE, tidy=TRUE, warning=FALSE}
knitr::kable(rbind(conUpTopTerms[conUpTopTerms$source == "GO:BP",][1,],
                   conUpTopTerms[conUpTopTerms$source == "REAC",][1,],
                   conUpTopTerms[conUpTopTerms$source == "WP",][1,]),
             format = "html")
```
**Table 9.** The top term from each of the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

Interestingly, there are 0 terms for the WP source. 

### Down-regulated
Now, do the same for the down-regulated genes.
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Down-regulated
conDownTopTerms <- gprofiler2::gost(query = rownames(conDownregulated), 
                                  organism = "hsapiens", 
                                  exclude_iea = TRUE,
                                  correction_method = "fdr",
                                  sources = c("GO:BP", "REAC", "WP"))

# Limit term size
conDownTopTerms <- data.frame(
  term_name = conDownTopTerms$result$term_name[conDownTopTerms$result$term_size < 500 &
                                               conDownTopTerms$result$term_size > 1],
  term_id = conDownTopTerms$result$term_id[conDownTopTerms$result$term_size < 500 &
                                           conDownTopTerms$result$term_size > 1],
  source = conDownTopTerms$result$source[conDownTopTerms$result$term_size < 500 &
                                         conDownTopTerms$result$term_size > 1]
)

# Visualize our results
knitr::kable(head(conDownTopTerms, 10), format = "html")
```
**Table 10.** The first top 10 terms across all the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 


Number of terms:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
length(conDownTopTerms$term_name)
```

We are going to want to look at the top terms for each source eventually, so we can look at it now. 
```{r message = FALSE, tidy=TRUE, warning=FALSE}
knitr::kable(rbind(conDownTopTerms[conDownTopTerms$source == "GO:BP",][1,],
                   conDownTopTerms[conDownTopTerms$source == "REAC",][1,],
                   conDownTopTerms[conDownTopTerms$source == "WP",][1,]),
             format = "html")
```
**Table 11.** The top term from each of the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

As we saw earlier from the up-regulated analysis for the CON samples, the WP source had 0 terms. We have a top term from WP in the down-regaulted genes as seen in Table 9. We can check this out.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
knitr::kable(head(conDownTopTerms[conDownTopTerms$source == "WP",], 10), format = "html")
nrow(conDownTopTerms[conDownTopTerms$source == "WP",])
```
**Table 12.** The top 10 terms from WP for the down-regulated genes in GM vs WM tissues in the CON patient group. 

### All DE genes
We will now be looking at all the differentially expressed genes as a whole. 
```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Up-regulated
conAllTopTerms <- gprofiler2::gost(query = rownames(results_con), 
                                  organism = "hsapiens", 
                                  exclude_iea = TRUE,
                                  correction_method = "fdr",
                                  sources = c("GO:BP", "REAC", "WP"))

# Limit term size
conAllTopTerms <- data.frame(
  term_name = conAllTopTerms$result$term_name[conAllTopTerms$result$term_size < 500 &
                                               conAllTopTerms$result$term_size > 1],
  term_id = conAllTopTerms$result$term_id[conAllTopTerms$result$term_size < 500 &
                                           conAllTopTerms$result$term_size > 1],
  source = conAllTopTerms$result$source[conAllTopTerms$result$term_size < 500 &
                                         conAllTopTerms$result$term_size > 1]
)

# Visualize our results
knitr::kable(head(conAllTopTerms, 10), format = "html")
```
**Table 13.** The first top 10 terms across all the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

Number of terms:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
length(conAllTopTerms$term_name)
```

Look for the top terms from each source:
```{r message = FALSE, tidy=TRUE, warning=FALSE}
knitr::kable(rbind(conAllTopTerms[conAllTopTerms$source == "GO:BP",][1,],
                   conAllTopTerms[conAllTopTerms$source == "REAC",][1,],
                   conAllTopTerms[conAllTopTerms$source == "WP",][1,]),
             format = "html")
```
**Table 14.** The top term from each of the three sources, GO Biological Process (GO:BP), Reactome (REAC), and WikiPathways (WP). The ID of each term is also included. 

## Thresholded over-representation analysis Discussion
### 1. Which method did you choose and why?

### 2. What annotation data did you use and why? What version of the annotation are you using?

### 3. How many genesets were returned with what thresholds?

### 4. Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)?

# Task 3: Interpretation
### 1. Do the over-representation results support conclusions or mechanism discussed in the original paper?

### 2. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your results.

# References
