---
title: "BCB420 Assignment 1: Data set selection and initial processing"
author: "Veronica Chang"
date: "February 14th, 2023"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    df_print: paged
editor_options: 
  markdown: 
    wrap: sentence
bibliography: a1.bib
---

# Task 1: Install packages and select an expression data set

## Packages
* [BiocManager](https://CRAN.R-project.org/package=BiocManager) [@morgan2022]
* [GEOmetadb](https://bioconductor.org/packages/release/bioc/html/GEOmetadb.html) [@zhu2008]
* [limma](https://bioconductor.org/packages/release/bioc/html/limma.html) [@smyth2015]
* [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) [@robinson2010]
* [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) [@durinck2009]
* [GEOquery](https://bioconductor.org/packages/release/bioc/html/GEOquery.html) [@davis2007]
* [Biobase](https://bioconductor.org/packages/release/bioc/html/Biobase.html) [@huber2015]
* [RColorBrewer](https://cran.r-project.org/package=RColorBrewer) [@neuwirth2022]
```{r, message=FALSE, warning=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")

library("BiocManager")
library("GEOmetadb")
library("limma")
library("edgeR")
library("dplyr")
library("Biobase")
library("biomaRt")
library("GEOquery")
library("RColorBrewer")
library("ggplot2")
library("knitr")
library("RSQLite")
library("kableExtra")
```

## Selected the dataset **GSE111972**

Description below [@vanderPoel2019]

```{r, message=FALSE, tidy=TRUE}
gseAccession <- "GSE111972"
gse <- GEOquery::getGEO(gseAccession, GSEMatrix = FALSE)
gpl <- names(GEOquery::GPLList(gse))[1]
gplInfo <- GEOquery::Meta(GEOquery::getGEO(gpl))
```

## Introduction

### Background
Multiple sclerosis (MS) is a chronic neuroinflammatory autoimmune disease. MS is characterized by demyelination and subsequent formation of lesions through out the central nervous system (CNS). Grey matter (GM) and white matter (WM) lesions in the brain show pathological differences. Microglia are CNS-resident and can function as antigen-presenting cells and phagocytes. Their role in MS is complex and controversial [@luo2017]. 

### General Experimental Design
In total 31 human microglia samples were analyzed from 10 MS donors (MS), including 5 grey matter and 10 white matter samples, and from 11 non-neurological control donors (CON), including 5 grey matter and 11 white matter samples. From each sample total RNA was extracted.[@vanderPoel2019]
	
### Study objectives
The authors isolated human microglia according to the previously described general experimental design. They suggest that pathological changes in MS tissue may be associated with changes in microglia, leading to lesion initiation. The authors wanted to analyze the transcriptional profile of microglia in the collected samples by RNA-sequencing [@luo2017]. Ultimately, they wanted to discover MS-related changes in microglia between the GM and WM brain regions.


# Task 2: Clean the data and map to HUGO symbols

## Fetch the data set

```{r, message=FALSE, tidy=TRUE}
gseDir <- "/home/rstudio/projects/A1"
if (! file.exists("~/projects/A1/GSE111972/GSE111972_norm_data.txt")) {
  sFiles = GEOquery::getGEOSuppFiles(gseAccession)
  fName = rownames(sFiles)
  GEOquery::gunzip(fNames[1])
}
```

There is only one supplemental file for this study.

```{r, tidy=TRUE}
file <- list.files(path = "~/projects/A1/GSE111972/", pattern = "\\.txt")
expData <- read.table(paste0("~/projects/A1/GSE111972/", file), sep = "\t", 
                        header = TRUE, check.names = FALSE)
colnames(expData)[2:32] <- sub("Sample_", "", colnames(expData))[2:32]
dim(expData)
```

There are 32 columns, the first being the Gene ID. There are 21283 rows (genes) in this data set.

## Define the groups

```{r, tidy=TRUE}
samples <- data.frame(lapply(colnames(expData)[2:32], 
        FUN=function(x){unlist(strsplit(x, 
                        split = "\\_"))[c(1,2,3)]}))
colnames(samples) <- sub("Sample_", "", colnames(expData))[2:32]
rownames(samples) <- c("patients_group","tissue","replicate")
samples <- data.frame(t(samples))

# made another group that combined patients_group and tissue for easier comparison
samples$group <- paste(samples$patients_group,samples$tissue)

samples
```

##Check for gene duplicates

```{r, tidy=TRUE}
length(unique(expData$ID))
!(length(unique(expData$ID)) == dim(expData)[1]) 
```

There are no gene duplicates. 

## Remove lowly expressed genes

```{r, message= FALSE, tidy=TRUE}
cpms = cpm(expData[,2:32])
rownames(cpms) <- expData[,1]

# get rid of low counts
keep = rowSums(cpms >1) >=10
expDataFiltered = expData[keep,]
```

## Identifier mapping

### Build a Biomart Query

```{r, message= FALSE, tidy=TRUE}
# get available marts
listMarts()
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
knitr::kable(head(datasets),format = "html")

# get only human
knitr::kable(head(datasets[grep(datasets$dataset,
                  pattern = "sapiens"),]),format = "html")
ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
dim(listFilters(ensembl))

knitr::kable(listFilters(ensembl)[1:10,1:2], type="html")
biomart_human_filters <- listFilters(ensembl)
```

### Attributes 

```{r, message= FALSE, tidy=TRUE}
kable(searchAttributes(mart = ensembl, 'hgnc') , format="html") %>%
  row_spec(2, background = "yellow")

httr::set_config(httr::config(ssl_verifypeer = FALSE))
ensembl <- biomaRt::useMart("ensembl", dataset="hsapiens_gene_ensembl")
```

### Check for genes without HGNC annotation

```{r, message= FALSE, tidy=TRUE}
conversion_stash <- "geneIDConversion.rds"
if (file.exists(conversion_stash)) {
  geneIDConversion <- readRDS(conversion_stash)
} else {
  geneIDConversion <- biomaRt::getBM(attributes = "hgnc_symbol",
                                  filters = "hgnc_symbol",
                                  values = expDataFiltered$ID,
                                  mart = ensembl)
  saveRDS(geneIDConversion, conversion_stash)
}

#find the genes that do not use HGNC annotation
not_hgnc <- expDataFiltered[-(which(rownames(keep) %in% geneIDConversion$hgnc_symbol)), ]
```

Seems like all genes are already annotated with HGNC symbols.

# Task 3: Apply Normalization

## Check statistical summaries

```{r, message = FALSE, tidy=TRUE, warning=FALSE}
summaryStats <- data.frame(apply(expDataFiltered[, 2:32], 2, summary))
summaryStats

# Calculate the summary statistics for each group

# CON (non-neurological control) WM (white matter)
summaryCONWM <- summary(rowMeans(expDataFiltered[, grep("CON_WM", colnames(expDataFiltered))]))
# MS (multiple sclerosis) WM (white matter)
summaryMSWM <- summary(rowMeans(expDataFiltered[, grep("MS_WM", colnames(expDataFiltered))]))
# CON (non-neurological control) GM (grey matter)
summaryCONGM <- summary(rowMeans(expDataFiltered[, grep("CON_GM", colnames(expDataFiltered))]))
# MS (multiple sclerosis) GM (grey matter)
summaryMSGM <- summary(rowMeans(expDataFiltered[, grep("MS_GM", colnames(expDataFiltered))]))

# Combine all
summaryStats2 <- rbind(summaryCONWM, summaryMSWM, summaryCONGM, summaryMSGM)
rownames(summaryStats2) <- c("CON WM", "MS WM", "CON GM", "MS GM")
knitr::kable(summaryStats2, format = "html", col.names = colnames(summaryStats2)) 
```

<br />

Means of the different groups are similar to each other for each tissue. This suggests that variations in gene counts can be attributed to biological variations rather than technical variations. 


## Check distribution

### Define colour groups for  plots
```{r, message = FALSE, tidy=TRUE, warning=FALSE}
# colour groups
tissueGroup <- as.factor(samples$tissue)
tissueGroup <- c("grey", "white")[tissueGroup]
patientGroup <- as.factor(samples$patients_group)
patientGroup <- c("forestgreen", "dodgerblue")[patientGroup]

colGroup <- as.factor(samples$group)
levels(colGroup) <-  brewer.pal(nlevels(colGroup), "Spectral")
colGroup <- as.character(colGroup)
```

### Plot a box plot
```{r, message = FALSE, tidy=TRUE, warning=FALSE}
# box plot
data2plot <- log2(cpm(expDataFiltered[,2:32]))
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM", 
        las = 2, cex = 0.5, cex.lab = 0.5, col = colGroup,
        cex.axis = 0.5, main = "Greyâ€“white matter of multiple sclerosis patients 
        and non-neurological control donors")

# draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)), 
       col = "red", lwd = 0.6, lty = "dashed")
```
<br />
**Figure 1. Box plot comparison differentiating between the different `patient_group` and `tissue`.**

The distribution and spread across all the samples are quite consistent.

### Plot a density plot. 

```{r, message = FALSE, tidy=TRUE, warning=FALSE}
# density plot
densityCounts <- apply(log2(edgeR::cpm(expDataFiltered[,3:8])),
                        2, density)

# calculate  limits
xlim <- 0; ylim <- 0
for (i in 1:length(densityCounts)) {
  xlim <- range(c(xlim, densityCounts[[i]]$x));
  ylim <- range(c(ylim, densityCounts[[i]]$y))
}

cols <- rainbow(length(densityCounts))
ltys <- rep(1, length(densityCounts))

# plot the first
plot(densityCounts[[1]], xlim=xlim, ylim=ylim, type="n",
     ylab="Smoothing density of log2-CPM",
     main="", cex.lab=0.85)

# plot the lines
for (i in 1:length(densityCounts))
  lines(densityCounts[[i]], col=cols[i], lty=ltys[i])

# add legend
legend("topright", colnames(data2plot),
       col=cols, lty=ltys, cex=0.75,
       border="blue", text.col = "black",
       merge = TRUE, bg = "snow2")

```
<br />
**Figure 2. Comparative distribution of gene expression across all the samples.**

The major peak between all the samples are similar. 

## Apply TMM normalization
```{r, message = FALSE, tidy=TRUE}
filteredData <-as.matrix(expDataFiltered[,2:32])
rownames(filteredData) <- expDataFiltered$ID
d <- edgeR::DGEList(counts=filteredData, group=samples$patients_group)
```

## Calculate the normalization factor
```{r, message = FALSE, tidy=TRUE}
d <- calcNormFactors(d)
expDataNorm <- cpm(d)
```

## Check distribution after normalization
### Boxplot

```{r warning=FALSE,message=FALSE}
data3plot <- log2(cpm(expDataNorm))
boxplot(data3plot, xlab = "Samples", ylab = "log2 CPM", 
        las = 2, cex = 0.5, cex.lab = 0.5, col = colGroup,
        cex.axis = 0.5, main = "Greyâ€“white matter of multiple sclerosis patients 
        and non-neurological control donors")

# draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)), 
       col = "red", lwd = 0.6, lty = "dashed")
```
<br />
**Figure 3. Box plot comparison differentiating between the different `patient_group` and `tissue` after TMM normalization.**

### Density plot

```{r, message = FALSE, tidy=TRUE, warning=FALSE}
# density plot
densityCounts2 <- apply(log2(edgeR::cpm(expDataNorm[,3:8])),
                        2, density)

# calculate  limits
xlim <- 0; ylim <- 0
for (i in 1:length(densityCounts2)) {
  xlim <- range(c(xlim, densityCounts2[[i]]$x));
  ylim <- range(c(ylim, densityCounts2[[i]]$y))
}

cols <- rainbow(length(densityCounts2))
ltys <- rep(1, length(densityCounts2))

# plot the first
plot(densityCounts2[[1]], xlim=xlim, ylim=ylim, type="n",
     ylab="Smoothing density of log2-CPM",
     main="", cex.lab=0.85)

# plot the lines
for (i in 1:length(densityCounts2))
  lines(densityCounts2[[i]], col=cols[i], lty=ltys[i])

# add legend
legend("topright", colnames(data3plot),
       col=cols, lty=ltys, cex=0.75,
       border="blue", text.col = "black",
       merge = TRUE, bg = "snow2")

```
<br />
**Figure 4. Comparative distribution of gene expression across all the samples after TMM normalization.**

There is not much of a difference between the normalized and the non-normalized data. Additionally, there is the same number of genes between them. 

## MDS plot

```{r eval=TRUE}
edgeR::plotMDS.DGEList(d, labels=rownames(samples),
  col = colGroup)
```

<br />
**Figure 3. MDS plot of all samples, coloured according to `group` which is the combination of `patient_group` and `tissue`.**

The tissues cluster together despite the patient group. 

### Final normalized data
```{r, tidy=TRUE, warning=FALSE}
kable(head(expDataNorm), format = "html")
```


# Task 4: Interpret, and document

### What are the control and test conditions of the dataset?
The control is the **CON** patient group that consisted of non-neurological control donors. This group is compared against the **MS** patient group which consisted of multiple sclerosis patients in two tissues: **GM** (grey matter) and **WM** (white matter) from the brain.

### Why is the dataset of interest to you?
I am currently conducting a research project for the course IMM450 where I am studying the sex bias in antigen-presenting cells in the mouse model experimental autoimmune encephalomyelitis for the human disease multiple sclerosis. While this research project is mainly wet lab, I want to try using bioinformatics to analyze differences in MS. 

### Were there expression values that were not unique for specific genes?
There were no duplicated in the gene IDs in the original data set. This stayed the same when trying to conduct identifier mapping. 

### How did you handle these?
N/A.

### Were there expression values that could not be mapped to current HUGO symbols?
No. 

### How many outliers were removed?
After filtering the lowly expressed genes, we were left with 15076 genes compared to the original 21283. A total of 6207 genes were removed. The number of genes did not change after normalization. 

### How did you handle replicates?
All replicates for each patient group and tissue were treated as separate samples. The replicates can be sorted by their `tissue` and their `patient_group`. In the MDS plot, we can see that the samples mostly cluster together as tissues. 
    
### What is the final coverage of your dataset?
The final dataset is made up of:

    - `r nrow(expDataNorm)` with unique identifiers.
    
    - 31 samples in total.
    
    - 2 patient groups *MS* and *CON*, and 2 tissues *GM*, *WM*.
    
    - 10 MS patients and 11 CON
    
    - 10 replicates for WM in the MS group and 11 replicates in the CON group.
    
    - 5 replicates for GM in both patient groups.
    
# References
